torch/multiprocessing/queue.py

```diff
@@ -31,11 +32,91 @@ def __getattr__(self, name):
 class Queue(multiprocessing.queues.Queue):
 
     def __init__(self, *args, **kwargs):
+        if hasattr(multiprocessing, "get_context"):
+            kwargs['ctx'] = multiprocessing.get_context()
         super(Queue, self).__init__(*args, **kwargs)
-        self._reader = ConnectionWrapper(self._reader)
-        self._writer = ConnectionWrapper(self._writer)
-        self._send = self._writer.send
-        self._recv = self._reader.recv
+
+        if sys.version_info < (3, 3):
+            self._reader = ConnectionWrapper(self._reader)
+            self._writer = ConnectionWrapper(self._writer)
+            self._send = self._writer.send
+            self._recv = self._reader.recv
+
+        self.sig_shutdown = multiprocessing.Value('i', 0)
+        self.put_lock = multiprocessing.RLock()
+        self.get_lock = multiprocessing.Lock()
+
+    def __setstate__(self, state):
+        default_state, extra_state = state
+        (self.sig_shutdown, self.put_lock, self.get_lock) = extra_state
+        super(Queue, self).__setstate__(default_state)
+
+    def __getstate__(self):
+        return (super(Queue, self).__getstate__(),
+                (self.sig_shutdown, self.put_lock, self.get_lock))
+
+    def shutdown(self):
+        if self.is_shutdown():
+            return
+        with self.sig_shutdown.get_lock():
+            with self.put_lock:
```

Comment:
Grab the put lock, put a None to try free the get lock, then signal shutdown. This makes sure no one is getting or putting when we try to shutdown, preventing a race condition on shutdown.